webhooks/services/billing_processor.py

```diff
@@ -0,0 +1,275 @@
+            )
+            await asyncio.sleep(retry_duration)  # Wait a bit before retrying
+
+    async def _process_billing_event(self, message: Dict[str, Any]) -> None:
+        """
+        Process billing events. Convert them to LAGO events and post them to LAGO
+        """
+        message_data = message.get("data")
+        if isinstance(message_data, bytes):
+            message_data = message_data.decode("utf-8")
+
+        groub_id, timestamp_ns_str = message_data.split(":")
+        timestamp_ns = int(timestamp_ns_str)
+
+        events = self.redis_service.get_billing_event(
+            groub_id, timestamp_ns, timestamp_ns
+        )
+
+        if not events:
+            # there are duplicate done event bc of acapy to cloudapi conversion
+            # the score get updated and the event is not found with old score
+            logger.debug(
+                f"No events found for group_id: {groub_id} and timestamp: {timestamp_ns}"
+            )
+            return
+
+        event: Dict[str, Any] = json.loads(events[0])
+
+        if event.get("topic") == "credentials":
+            # handel credentials event
+            lago = self._convert_credential_event(event)
+
+        if event.get("topic") == "proofs":
+            # handel proofs event
+            lago = self._convert_proofs_event(event)
+
+        if event.get("topic") == "endorsements":
+            # handel endorsements event
+            lago = self._convert_endorsements_event(event)
+
+        if event.get("topic") == "issuer_cred_rev":
+            # handel issuer_cred_rev event
+            lago = self._convert_issuer_cred_rev_event(event)
+
+        # post billing event to LAGO
+        await self._post_billing_event(lago)
+
+    async def _post_billing_event(self, event: BillingEvent) -> None:
+        """
+        Post billing event to LAGO
+        """
+        logger.debug(f"Posting billing event: {event}")
+        try:
+            lago_response = await self._client.post(
+                url=self._LAGO_URL,
+                json={"event": event.model_dump()},
+            )
+
+            logger.error(f"Response from LAGO: {lago_response.json()}")
+            lago_response.raise_for_status()
+
+        except HTTPException as e:
+            if e.status_code == 422:
+                logger.error(f"Error posting billing event >>> : {e.detail}")
+                pass
+            logger.error(f"Error posting billing event: {e}")
+            raise e
+
+    def _convert_credential_event(
+        self, event: Dict[str, Any]
+    ) -> CredentialBillingEvent:
+        """
+        Convert credential event to LAGO event
+        """
+        logger.debug(f"Converting credential event: {event}")
+
+        # using thread_id as transaction_id
+        payload: Dict[str, Any] = event.get("payload")
+        lago_event = CredentialBillingEvent(
+            transaction_id=payload.get("thread_id"),
+            external_customer_id=event.get("group_id"),
```

Comment:
Keep in mind, these CAN be None ... If Lago handles null values, it's fine. Otherwise we might want to add `event.get("group_id") or "no group"` or something so it has a default empty value. Not sure if it's necessary